object-model/copy-semantics.md

```diff
@@ -0,0 +1,153 @@
+# C++ object model: copy semantics
+_Skeleton instructions are typeset in italic text._
+
+## Overview
+
+_Provides a short natural language abstract of the module’s contents._
+_Specifies the different levels of teaching._
+
+<table>
+  <thead>
+    <th>Level</th>
+    <th>Objectives</th>
+  </thead>
+  <tr>
+    <td>Foundational knowledge</td>
+    <td>understanding how and when are copies made</td>
+  </tr>
+  <tr>
+    <td>Advanced</td>
+    <td>implementing user-defined copy operations</td>
+  </tr>
+  <tr>
+    <td>Further studies</td>
+    <td>special cases: copy elision</td>
+  </tr>
+</table>
+
+## Motivation
+
+_Why is this important?_
+_Why do we want to learn/teach this topic?_
+
+Copy semantics allows the user to determine how objects of a class get replicated and interact on a value level.
+
+## Topic introduction
+
+_Very brief introduction to the topic._
+
+Explains when and how objects are copied.
+
+## Foundational knowledge: How and when are copies made
+
+### Background/Required Knowledge
+
+A student is able to:
+* explain what a C++ type is?               [[C++ object model: types]][1]
+* explain what an object is?                [[C++ object model: objects]][2], [[C++ object model: constant objects]][3]
+* define and understand class invariants?
+
+It helps when a student is able to:
+* use move semantics                        [[C++ object model: move semantics]][4]
+* explain special member functions          [[C++ object model: special member functions]][5]
+
+### Student outcomes
+
+_A list of things "a student should be able to" after the curriculum._
+_The next word should be an action word and testable in an exam._
+_Max 5 items._
+
+A student should be able to:
+
+1. explain what copy semantics accomplish
+  * establishing "equivalent" object state in another object
+2. explain difference between copying a reference and copying a value\*
+3. explain where copies are made
+
+\* In other languages these differences are sometimes referred to as shallow and deep copy.
+
+### Caveats
+
+_This section mentions subtle points to understand, like anything resulting in
+implementation-defined, unspecified, or undefined behavior._
+
+* Compiler-provided (reference based) copies may result in ownership problems (e.g., `char*`).
+
+### Points to cover
+
+_This section lists important details for each point._
+
+* Principle of copying
+  * Copying of trivially copyable types
+  * Copying of non-trivial class types
+  * Copying an object does not change the original
+* Copy Flavors
+  * Copy by value
+  * Copy by reference
```

Comment:
@vulder to find a document where "flavors" were discussed in Prague(?), and link it here. Anyway, this comment thread needn't block merging this PR as far as I'm concerned.

Reply:
Here is the document from prague, which is basically the raw version of the PR
https://docs.google.com/document/d/1xmfxLMIgQUPCUKARDJDfPYzLwGniPlzlNtzZtF2XdXs/edit?usp=sharing

Reply:
If you like alternative wording for `Flavors`, fine with me. Any ideas?

Reply:
Ah, I see the history now. I bet:
- this used to say "Deep copy vs. shallow copy"
- I (and others) objected to teaching any distinction between "deep" and "shallow" (as opposed to Just Copy vs. everything else in the world none of which is semantically the same as copying)
- The objection was heard as if I was objecting to the _words_ "deep" and "shallow"
- The words were changed to "by value" and "by reference"
- ...which just confuses the matter at an even deeper level.

I continue to recommend just striking these bullet points. The way you write a copy operation is, _first_ you decide what your type means by "copy," and _then_ you implement those semantics under the "copy constructor" syntax. There may be superficially different kinds of _types_ — a `FooHandle` type might have handle semantics while a `FooHolder` type might have value semantics — but the meaning of "copy" is the same in both cases; it means "take whatever this thing represents, its _value_ — whether that's a foo or a handle-to-a-foo or an integer or a pointer or a mutex-lock or whatever — and make this value look like that one." No deep, no shallow; just "do what's needed so that after this operation A will represent the same value as B."

Reply:
Yep. Fine with me.